src/Jira/oauth/OnPremOAuth.ts

```diff
@@ -1,239 +1,243 @@
-        return [...Array(nonceSize)].map(() => NONCE_CHARS[Math.floor(Math.random() * NONCE_CHARS.length)]).join('');
-    }
-
-    private static encodeData(toEncode: string): string {
-        return encodeURIComponent(toEncode).replace(/!/g, "%21")
-                        .replace(/'/g, "%27")
-                        .replace(/\(/g, "%28")
-                        .replace(/\)/g, "%29")
-                        .replace(/\*/g, "%2A");
-    }
-
-    private static normalizeUrl(url: string|URL): string {
-        url = typeof url === "string" ? new URL(url) : url;
-        let port = "";
-        if (url.port) { 
-            if ((url.protocol == "http:" && url.port != "80" ) ||
-                (url.protocol == "https:" && url.port != "443") ) {
-                    port = ":" + url.port;
-                }
+export const assertOAuthRequestToken = (body: Record<string, string>): OAuthBody => {
+    try {
+        if (typeof body.oauth_token !== "string" || !body.oauth_token) {
+            throw Error("Unexpected OAuth response from server: missing or invalid oauth_token");
         }
-        if (!url.pathname || url.pathname == "") {
-            url.pathname = "/";
+        if (typeof body.oauth_token_secret !== "string" || !body.oauth_token_secret) {
+            throw Error("Unexpected OAuth response from server: missing or invalid oauth_token_secret");
+        }
+    } catch (error) {
+        log.info(`Unexpected response from JIRA:`, JSON.stringify(body));
+        throw error;
+    }
+    return {
+        oauth_token: body.oauth_token,
+        oauth_token_secret: body.oauth_token_secret,
+    };
+};
+
+export const buildAuthorizationHeaders = (orderedParameters: [string, string][]) => {
+    // Whilst the all the parameters should be included within the signature, only the oauth_ arguments
+    // should appear within the authorization header.
+    const authParams = orderedParameters.filter(([key]) => isParameterNameAnOAuthParameter(key));
+    // Convert to strings: ["oauth_foo", "bar"] becomes oauth_foo="bar"
+    const authStrings = authParams.map(([key, value]) => `${encodeData(key)}="${encodeData(value)}"`);
+    return `OAuth ${authStrings.join(",")}`;
+}
+
+export const createSignatureBase = (method: string, url: string, parameters: string): string => {
+    return `${method.toUpperCase()}&${encodeData(normalizeUrl(url))}&${encodeData(parameters)}`;
+}
+
+export const encodeData = (toEncode: string): string => {
+    return encodeURIComponent(toEncode).replace(/!/g, "%21")
+        .replace(/'/g, "%27")
+        .replace(/\(/g, "%28")
+        .replace(/\)/g, "%29")
+        .replace(/\*/g, "%2A");
+}
+
+/**
+ * Is the parameter considered an OAuth parameter?
+ */
+const isParameterNameAnOAuthParameter = (parameter: string): boolean => {
+    return parameter.startsWith("oauth_");
+}
+
+export const makeArrayOfArgumentsHash = (argumentsHash: Map<string, string | string[]>): [string, string][] => {
```

Comment:
Not sure why this is `Map<string, string | string[]>` instead of `Map<string, string>`.
Issue #621

This was an issue before and I didn't remove the support for `string[]`.
Please let's not make that a blocker for this PR.